crypto/BRCryptoKey.c

```diff
@@ -0,0 +1,278 @@
+    return EMPTY_ADDRESS_PARAMS;
+}
+
+extern BRCryptoKey
+cryptoKeyCreateFromStringPrivate (const char *string) {
+    BRAddressParams params = cryptoKeyFindAddressParams(string);
+
+    BRKey core;
+
+    return (1 == BRKeySetPrivKey (&core, params, string)
+            ? cryptoKeyCreateInternal (core, params)
+            : NULL);
+}
+
+extern BRCryptoKey
+cryptoKeyCreateForPigeon (BRCryptoKey privateKey, uint8_t *nonce, size_t nonceCount) {
+    BRKey pairingKey;
+
+    BRKeyPigeonPairingKey (&privateKey->core, &pairingKey, nonce, nonceCount);
+
+    return cryptoKeyCreateInternal (pairingKey, CRYPTO_ADDRESS_PARAMS);
+}
+
+extern BRCryptoKey
+cryptoKeyCreateForBIP32ApiAuth (const char *phrase, const char *words[]) {
+    if (!BRBIP39PhraseIsValid (words, phrase)) return NULL;
+
+
+    UInt512 seed;
+    BRBIP39DeriveKey (seed.u8, phrase, NULL);
+
+    BRKey core;
+    BRBIP32APIAuthKey (&core, &seed, sizeof (UInt512));
+
+
+    //        BRBIP39DeriveKey(&seed, phrase, nil)
+    //        BRBIP32APIAuthKey(&key, &seed, MemoryLayout<UInt512>.size)
+    //        seed = UInt512() // clear seed
+    //        let pkLen = BRKeyPrivKey(&key, nil, 0)
+    //        var pkData = CFDataCreateMutable(secureAllocator, pkLen) as Data
+    //        pkData.count = pkLen
+    //        guard pkData.withUnsafeMutableBytes({ BRKeyPrivKey(&key, $0.baseAddress?.assumingMemoryBound(to: Int8.self), pkLen) }) == pkLen else { return nil }
+    //        key.clean()
+
+    return cryptoKeyCreateInternal(core, CRYPTO_ADDRESS_PARAMS);
+}
+
+extern BRCryptoKey
+cryptoKeyCreateForBIP32BitID (const char *phrase, int index, const char *uri,  const char *words[]) {
+    if (!BRBIP39PhraseIsValid (words, phrase)) return NULL;
+
+
+    UInt512 seed;
+    BRBIP39DeriveKey (seed.u8, phrase, NULL);
+
+    BRKey core;
+    BRBIP32BitIDKey (&core, &seed, sizeof(UInt512), index, uri);
+
+    return cryptoKeyCreateInternal(core, CRYPTO_ADDRESS_PARAMS);
+}
+
+typedef enum {
+    SERIALIZE_PUBLIC_IDENTIFIER = 1,
+    SERIALIZE_PRIVATE_IDENTIFIER = 2
+} BRCryptoKeySerializeIdentifier;
+
+
+extern BRCryptoKey
+cryptoKeyCreateFromSerializationPublic (uint8_t *data, size_t dataCount) {
+    if (dataCount < 1) return NULL;
+    if (SERIALIZE_PUBLIC_IDENTIFIER != (BRCryptoKeySerializeIdentifier) data[0]) return NULL;
+
+    BRKey core;
+
+    return (1 == BRKeySetPubKey (&core, &data[1], dataCount - 1)
+            ? cryptoKeyCreateInternal (core, CRYPTO_ADDRESS_PARAMS)
+            : NULL);
+}
+
+extern size_t
+cryptoKeySerializePublic (BRCryptoKey key, /* ... */ uint8_t *data, size_t dataCount) {
+    size_t publicKeySize = BRKeyPubKey (&key->core, NULL, 0);
+    size_t serializeSize = 1 + publicKeySize;
+
+    if (NULL == data) return serializeSize;
+    if (dataCount < serializeSize) return 0;
+
+    data[0] = SERIALIZE_PUBLIC_IDENTIFIER;
+    BRKeyPubKey (&key->core, &data[1], dataCount);
```

Comment:
dataCount - 1?

Reply:
The serialization code is no longer used, at this time.  Maybe best to remove it from the C.  Let me address some of Ehsan comments and see where it goes.